File: python/dynare_py/steady_state.py
"""
Steady-state utilities.

For a log-linearised model the log-deviation steady state is always y* = 0.
This module provides tools for the *nonlinear* steady state — useful when
deriving the log-linear approximation or verifying parameters.
"""

from __future__ import annotations

from typing import Callable, Dict, List, Optional, Sequence

import numpy as np
from scipy.optimize import fsolve, root


def solve_steady_state(
    equations: Callable[[np.ndarray], np.ndarray],
    x0: np.ndarray,
    method: str = "hybr",
    tol: float = 1e-10,
    verbose: bool = False,
) -> Dict:
    """
    Solve the nonlinear steady-state system f(x) = 0.

    Parameters
    ----------
    equations : callable
        Function f(x) -> residuals.  Must return array of same size as x.
    x0 : initial guess
    method : scipy.optimize.root method ('hybr', 'lm', 'broyden1', ...)
    tol : convergence tolerance
    verbose : print solver output

    Returns
    -------
    dict with keys 'x' (solution), 'success', 'residual', 'message'
    """
    sol = root(equations, x0, method=method, tol=tol,
               options={"disp": verbose})
    return {
        "x": sol.x,
        "success": sol.success,
        "residual": float(np.max(np.abs(sol.fun))),
        "message": sol.message,
    }


def log_linearise(
    equations: Callable[[np.ndarray], np.ndarray],
    ss: np.ndarray,
    eps: float = 1e-5,
) -> np.ndarray:
    """
    Numerically compute the Jacobian of `equations` at steady state `ss`.

    This is the log-linear coefficient matrix when equations are written in
    terms of log-deviations and ss contains the *level* steady-state values.
    Each column j is (f(ss with x_j perturbed) - f(ss)) / eps,
    scaled by ss[j] to give log-derivative.

    Returns
    -------
    J : (n_eq, n_vars) Jacobian (d f / d log x_j * x_j)
    """
    n = len(ss)
    f0 = equations(ss)
    n_eq = len(f0)
    J = np.zeros((n_eq, n))
    for j in range(n):
        x_plus = ss.copy()
        x_plus[j] = ss[j] * (1 + eps)
        J[:, j] = (equations(x_plus) - f0) / eps
    return J


def check_steady_state(
    equations: Callable[[np.ndarray], np.ndarray],
    ss: np.ndarray,
    tol: float = 1e-8,
) -> bool:
    """Return True if max |f(ss)| < tol."""
    residual = np.max(np.abs(equations(ss)))
    return bool(residual < tol)

File: julia/DynareJL/src/steady_state.jl
"""
Steady-state utilities for DynareJL.

For log-linearised models the log-deviation steady state is y* = 0.
This module handles the nonlinear steady state for model derivation/verification.
"""

"""
    solve_steady_state(f!, x0; tol=1e-10, maxiter=500) -> NamedTuple

Find x s.t. f!(x) ≈ 0 using Newton iterations with backtracking.

# Arguments
- `f!` : in-place function f!(residual, x) or out-of-place f(x)
- `x0` : initial guess
"""
function solve_steady_state(f, x0::AbstractVector; tol=1e-10, maxiter=500)
    x = copy(Float64.(x0))
    n = length(x)

    for iter in 1:maxiter
        F = f(x)
        err = maximum(abs, F)
        err < tol && return (x=x, success=true, residual=err, iters=iter)

        # Numerical Jacobian
        J = _numerical_jacobian(f, x)
        dx = -J \ F

        # Backtracking line search
        step = 1.0
        for _ in 1:20
            xnew = x .+ step .* dx
            Fnew = f(xnew)
            norm(Fnew) < norm(F) && (x = xnew; break)
            step *= 0.5
        end
    end

    F = f(x)
    return (x=x, success=false, residual=maximum(abs, F), iters=maxiter)
end

function _numerical_jacobian(f, x; eps=1e-6)
    n  = length(x)
    F0 = f(x)
    J  = zeros(length(F0), n)
    for j in 1:n
        xp    = copy(x); xp[j] += eps
        J[:,j] = (f(xp) .- F0) ./ eps
    end
    return J
end

"""
    log_linearise(f, ss; eps=1e-5) -> Matrix{Float64}

Numerically compute the log-linearised Jacobian at steady state `ss`.
Entry (i,j) = ∂f_i/∂log(x_j) * x_j at x=ss.
"""
function log_linearise(f, ss::AbstractVector; eps=1e-5)
    n   = length(ss)
    F0  = f(ss)
    J   = zeros(length(F0), n)
    for j in 1:n
        xp      = copy(Float64.(ss)); xp[j] *= (1 + eps)
        J[:,j]  = (f(xp) .- F0) ./ eps
    end
    return J
end

"""
    check_steady_state(f, ss; tol=1e-8) -> Bool
"""
check_steady_state(f, ss; tol=1e-8) = maximum(abs, f(ss)) < tol
